### PR TITLE
[Enhancement] Push distinct agg down window operator

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes # https://github.com/StarRocks/starrocks/issues/19134
### Description
Push Distinct Agg down below WindowOperators, for an example:

```SQL
-- Q1
select distinct a, b, c, sum(d) over(partition by a, b order by c) from t;
-- Q2
select distinct a, b, c, sum(d) over(partition by a, b order by c) from (
select a,b,c,sum(d) as d from t group by a,b,c) t
```
Q1 is equivalent to Q2, however Q2 is far more efficient than Q1 when a, b and c columns are low-cardinality and there are many duplicates of (a,b,c) in table t,  since group-by aggregation below WindowOperators reduces the number of rows processed by WindowOperators drastically.

This PR add PushDownDistinctAggregationRewriter  to transform Q1 into Q2.

### Test
Test is conducted on SR cluster:   3 x BE(Cpu: 16 cores; Mem: 64GB)
Dataset: 1.7 billion rows, skewRation=69%, distinct_cardinality(region,pay_date)=7409

Test queries as folows:
- Q1: original SQL
- Q2: manual rewritten SQL of Q1

```SQL
-- Q1
SELECT DISTINCT
COALESCE(region,'Others') AS region,
pay_date,
SUM(income) OVER ( PARTITION BY COALESCE(region,'Others'),pay_date) AS gp_income,
SUM(income) OVER ( PARTITION BY COALESCE(region,'Others'),MONTH(pay_date) ORDER BY pay_date) AS gp_income_MTD,
SUM(income) OVER ( PARTITION BY COALESCE(region,'Others'),YEAR (pay_date),QUARTER(pay_date) ORDER BY pay_date) AS gp_income_QTD,
SUM(income) OVER ( PARTITION BY COALESCE(region,'Others'),YEAR (pay_date) ORDER BY pay_date) AS gp_income_YTD
FROM trans
where month(pay_date)=1
order by region,pay_date

-- Q2
select region,pay_date,
SUM(income) OVER ( PARTITION BY region,pay_date) AS gp_income,
SUM(income) OVER ( PARTITION BY region,MONTH(pay_date) ORDER BY pay_date) AS gp_income_MTD,
SUM(income) OVER ( PARTITION BY region,YEAR (pay_date),QUARTER(pay_date) ORDER BY pay_date) AS gp_income_QTD,
SUM(income) OVER ( PARTITION BY region,YEAR (pay_date) ORDER BY pay_date) AS gp_income_YTD
from (
select  COALESCE(region,'Others') AS region,
pay_date,
sum(income) as income
from trans
where month(pay_date) = 1
group by region,pay_date
) t1
order by region,pay_date
```
**Test result**

Q1(-opt) means Q1 without optimzation, Q1(+opt) means with optimization, Q2(manual) means manual rewritten sql from Q1.  For end-to-end response time, Q1(+opt) outperforms Q1(+opt) 120X,  Q1(+opt) uses 1/63 PeakMemoryUsage as Q1(-opt).  Q1(+opt) is verified aproximately as good as Q2(manual).
```
+=============+==========+=================+
| test query  | RT       | PeakMemoryUsage |
+=============+==========+=================+
| Q1(-opt)    | 265.227s | 25.48GB         |
| Q1(+opt)    | 2.208s   | 409.79MB        |
| Q2(manual)  | 2.11s    | 320.10MB        |
+-------------+----------+-----------------+
| Improvement | 120X     | 63X             |
+-------------+----------+-----------------+
```

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Comment by satanson** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+        public AggregatePushDownContext visitLogicalFilter(OptExpression optExpression,
+                                                           AggregatePushDownContext context) {
+            if (canNotPushDown(optExpression, context)) {
+                return visit(optExpression, context);
+            }
+
+            // add filter columns in groupBys
+            LogicalFilterOperator filter = optExpression.getOp().cast();
+            filter.getRequiredChildInputColumns().getStream().mapToObj(factory::getColumnRef)
+                    .forEach(v -> context.groupBys.put(v, v));
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:263]: Why add filter columns into `groupBys`, this filter is above the window or below the window?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 2]
```diff
@@ -329,6 +330,7 @@ private OptExpression rewriteAndValidatePlan(OptExpression tree, TaskContext roo
 
     private OptExpression pushDownAggregation(OptExpression tree, TaskContext rootTaskContext,
                                               ColumnRefSet requiredColumns) {
+        tree = new PushDownDistinctAggregateRule().rewrite(tree, rootTaskContext);
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:333]: I suggest put this line below the config? so we can disable aggregate push down by using session vars.

```
if (context.getSessionVariable().getCboPushDownAggregateMode() == -1) {
            return tree;
}

 tree = new PushDownDistinctAggregateRule().rewrite(tree, rootTaskContext);

```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+
+        @Override
+        public RewriteInfo visitLogicalProject(OptExpression optExpression, RewriteInfo rewriteInfo) {
+            if (!rewriteInfo.getRemapping().isPresent()) {
+                return RewriteInfo.NOT_REWRITE;
+            }
+            ReplaceColumnRefRewriter replacer = rewriteInfo.getRemapping().get().getReplacer();
+            LogicalProjectOperator projectOp = optExpression.getOp().cast();
+            ColumnRefSet columnRefSet = new ColumnRefSet();
+            columnRefSet.union(getReferencedColumnRef(new ArrayList<>(projectOp.getColumnRefMap().keySet())));
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:354]: why put the keyset in a arrayList?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalWindowOperator.java [hunk 4]
```diff
@@ -73,6 +76,18 @@ public ColumnRefSet getOutputColumns(ExpressionContext expressionContext) {
         }
     }
 
+    public boolean canPushDownAgg() {
+        //TODO(by satanson): support AVG and COUNT in future
+        Set<String> supportWindowFun = Sets.newHashSet(FunctionSet.SUM);
+        AnalyticWindow window = this.getAnalyticWindow();
+        return window == null || window.getType().equals(AnalyticWindow.Type.RANGE) &&
+                !window.getLeftBoundary().getType().equals(AnalyticWindow.BoundaryType.PRECEDING) &&
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalWindowOperator.java:84]: I think it must not be equal to `CURRENT_ROW` too

Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalWindowOperator.java:84]: why we limit `!window.getLeftBoundary().getType().equals(AnalyticWindow.BoundaryType.PRECEDING)`

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java (hunk 3 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:354]: fixed

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java (hunk 1 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:263]: commented added

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalWindowOperator.java [hunk 5]
```diff
@@ -73,6 +76,18 @@ public ColumnRefSet getOutputColumns(ExpressionContext expressionContext) {
         }
     }
 
+    public boolean canPushDownAgg() {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalWindowOperator.java:79]: move it to rewrite rule

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+
+        public ColumnRefRemapping() {
+            remapping = Maps.newHashMap();
+        }
+
+        public ColumnRefRemapping(Map<ColumnRefOperator, ColumnRefOperator> remapping) {
+            this.remapping = remapping;
+        }
+
+        void combine(ColumnRefRemapping other) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:79]: private/public

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+            remapping.putAll(other.remapping);
+            cachedReplacer = Optional.empty();
+            cachedColumnRefSet = Optional.empty();
+        }
+
+        ReplaceColumnRefRewriter getReplacer() {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:85]: private/public

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+            if (!cachedReplacer.isPresent()) {
+                ReplaceColumnRefRewriter replacer =
+                        new ReplaceColumnRefRewriter(remapping.entrySet().stream().collect(
+                                Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)), true);
+                cachedReplacer = Optional.of(replacer);
+            }
+            return cachedReplacer.get();
+        }
+
+        ColumnRefSet getColumnRefSet() {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:95]: private/public

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+            if (!cachedColumnRefSet.isPresent()) {
+                cachedColumnRefSet = Optional.of(new ColumnRefSet(remapping.keySet()));
+            }
+            return cachedColumnRefSet.get();
+        }
+
+        boolean isEmpty() {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:102]: private/public

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java [hunk 10]
```diff
@@ -0,0 +1,458 @@
...
+            return remapping.isEmpty();
+        }
+    }
+
+    public static class RewriteInfo {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:107]:
```suggestion
    private static class RewriteInfo {
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+            return ctx;
+        }
+
+        public void setCtx(AggregatePushDownContext ctx) {
+            this.ctx = ctx;
+        }
+
+        private AggregatePushDownContext ctx;
+
+        public static RewriteInfo NOT_REWRITE = new RewriteInfo(false, null, null, null);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:122]:
```suggestion
        public final static RewriteInfo NOT_REWRITE = new RewriteInfo(false, null, null, null);
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+
+            LogicalScanOperator scanOp = optExpression.getOp().cast();
+            ColumnRefSet scanOutputColRefSet = new ColumnRefSet(scanOp.getOutputColumns());
+            Map<CallOperator, ColumnRefOperator> uniqueAggregations = Maps.newHashMap();
+            Map<ColumnRefOperator, ColumnRefOperator> remapping = Maps.newHashMap();
+            Preconditions.checkArgument(scanOutputColRefSet.containsAll(new ColumnRefSet(groupBys)));
+
+            for (Map.Entry<ColumnRefOperator, CallOperator> entry : ctx.aggregations.entrySet()) {
+                CallOperator aggCall = entry.getValue();
+                Preconditions.checkArgument(aggCall.getChildren().size() == 1 && aggCall.getChild(0).isColumnRef());
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:326]: I think you can't promise the check always true

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+            }
+
+            LogicalProjectOperator projectOp = optExpression.getOp().cast();
+
+            if (projectOp.getColumnRefMap().entrySet().stream().allMatch(e -> e.getValue().equals(e.getKey()))) {
+                return context;
+            }
+
+            // rewrite
+            ReplaceColumnRefRewriter rewriter = new ReplaceColumnRefRewriter(projectOp.getColumnRefMap());
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:237]: It seems there is special process for `case when` and `if` fucntion in `PushDownAggregateRewriter.rewriteProject()`. Should we consider it here?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:333]: not required,  one day these two rules would be fused into one,  for now, PushDownDistinctAggregateRule only process distinct-agg-window queries,  PushDownAggregateRule only process agg-join queries.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java (hunk 13 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:237]: case when and if function will be processed in the future, but it is quite different from PushDownAggregateRewriter's present way.  A right way is to allow push-down agg to sits above union/project/join operator.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+        public void setRewritten(boolean rewritten) {
+            this.rewritten = rewritten;
+        }
+
+        public Optional<ColumnRefRemapping> getRemapping() {
+            if (!rewritten || remapping.isEmpty()) {
+                return Optional.empty();
+            }
+            return Optional.of(remapping);
+        }
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:145]: I think don't need Optional method and may be easier if add a method `hasRemapping`?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalWindowOperator.java (hunk 4 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalWindowOperator.java:84]: the limit is not required

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java (hunk 12 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:326]: at present,  only sum(column) over (...) is processed, so the promise can be ensured.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java (hunk 14 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownDistinctAggregateRewriter.java:145]: Optional object is more convenient, we can invoke orElse, ifPresent, map method.

**Review by Seaven — DISMISSED** [review]

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Pull request closed without merging**